File: grammar.py
# -*- coding: utf-8 -*-
"""
Created on Thu Oct 31 19:34:26 2019
@author: Nathan Srirama
This is my code for "parsing" a valid sentence as defined below using a 
finite state machine (I can't remember if this is a Moore or Mealy version)

Current state | Next word | Next state
--------------|-----------|-----------
    S0        |    the    |     S1
    S1        |    lazy   |     S2
    S1        |    smelly |     S2
    S1        |    dog    |     S3
    S1        |    cat    |     S3
    S2        |    lazy   |     S2
    S2        |    smelly |     S2
    S2        |    cat    |     S3
    S2        |    dog    |     S3
    S3        |    ran    |     S4
    S3        |    ate    |     S4
    S4        |    slowly |     S5
    S4        |    noisily|     S5
    S4        |    EOI    |     End
    S5        |    EOI    |     End
"""

states = [
          [0, "the", 1],
          [1, "lazy", 2],
          [1, "smelly", 2],
          [1, "dog", 3],
          [1, "cat", 3],
          [2, "lazy", 2],
          [2, "smelly", 2],
          [2, "cat", 3],
          [2, "dog", 3],
          [3, "ran", 4],
          [3, "ate", 4],
          [4, "slowly", 5],
          [4, "noisily", 5],
          [4, "", -1],
          [5, "", -1]
         ]

def main():
    testFile = open("tests.txt")
    sentences = [x.strip() for x in testFile]
    for string_to_parse in sentences:
        flag = isValidSentence(string_to_parse.lower())
        if flag:
            print("\"" + string_to_parse + "\" is a valid sentence!")
        else:
            print("\"" + string_to_parse + "\" is not a valid sentence -_-")


def isValidSentence(string_to_parse):
    #sets the current state to the starting state
    curr_state = 0
    data = string_to_parse.split()
    #I append an empty string so that the function will know what a valid 
    #sentence end is vs. a premature end. I like this better than putting EOI
    #at the end of the sentences, which looks weird
    data.append("")
    for word in data:
        found = False
        #I wanted to try a find function of some sort to make this not so 
        #brute-force-ish, but my python knowledge is meh and I didn't know how
        #find functions would work with my list of lists used for states
        for state in states:
            if state[0] == curr_state and state[1] == word:
                curr_state = state[2]
                found = True
                break
        if not found:
            return False
    if curr_state == -1:    
        return True
    else:
        return False


    
if __name__ == "__main__":
    main()

File: grammar.cpp
#include <vector>
#include <iostream>

static bool interpreter(std::vector<std::string> sentence) {
	bool isValid = true;
	int node = 0;
	int curr = 0;
	while (curr < sentence.size()) {
		if (node == 0 && sentence[curr] == "the") {
			node = 1;
			curr++;
		}
		else if (node == 1 && (sentence[curr] == "lazy" || sentence[curr] == "smelly")) {
			node = 3;
			curr++;
		}
		else if (node == 1 && (sentence[curr] == "dog" || sentence[curr] == "cat")) {
			node = 2;
			curr++;
		}
		else if (node == 3 && (sentence[curr] == "lazy" || sentence[curr] == "smelly")) {
			curr++;
		}
		else if (node == 3 && (sentence[curr] == "dog" || sentence[curr] == "cat")) {
			node = 2;
			curr++;
		}
		else if (node == 2 && (sentence[curr] == "ate" || sentence[curr] == "ran")) {
			node = 4;
			curr++;
		}
		else if (node == 4 && (sentence[curr] == "slowly" || sentence[curr] == "noisily")) {
			node = 5;
			curr++;
		}
		else if (node == 4 && (sentence[curr] == "EOI")) {
			return true;
		}
		else if (node == 5 && sentence[curr] == "EOI") {
			return true;
		}
		else { return false; }

	}
	isValid = false;
	return isValid;

}



int main() {
	//check passing
	std::vector<std::string> shouldPass1 = { "the", "cat", "ate", "EOI" };
	std::vector<std::string> shouldPass2 = { "the", "lazy", "dog", "ate", "EOI" };
	std::vector<std::string> shouldPass3 = { "the", "cat", "ran", "noisily", "EOI" };
	std::vector<std::string> shouldPass4 = { "the", "lazy", "smelly", "lazy", "smelly", "cat", "ate", "EOI" };
	bool isValid = interpreter(shouldPass1);
	if (isValid) {
		std::cout << "good 1" << std::endl;
	}
	else {
		std::cout << "fail 1" << std::endl;
	}

	isValid = interpreter(shouldPass2);
	if (isValid) {
		std::cout << "good 2" << std::endl;
	}
	else {
		std::cout << "fail 2" << std::endl;
	}

	isValid = interpreter(shouldPass3);
	if (isValid) {
		std::cout << "good 3" << std::endl;
	}
	else {
		std::cout << "fail 3" << std::endl;
	}

	isValid = interpreter(shouldPass4);
	if (isValid) {
		std::cout << "good 4" << std::endl;
	}
	else {
		std::cout << "fail 4" << std::endl;
	}


	//check failing
	std::vector<std::string> shouldFail1 = { "the", "ate", "EOI" };
	std::vector<std::string> shouldFail2 = { "the", "lazy", "dog", "cat", "ate", "EOI" };
	std::vector<std::string> shouldFail3 = { "the", "cat", "ran", "noisily" };
	std::vector<std::string> shouldFail4 = { "the", "lazy", "EOI" };
	std::vector<std::string> shouldFail5 = { "the", "EOI" };
	std::vector<std::string> shouldFail6 = { "lazy", "dog", "cat", "ate", "EOI" };
	std::vector<std::string> shouldFail7 = { "the", "cat", "ran", "slowly", "noisily" };
	std::vector<std::string> shouldFail8 = { "the", "lazy", "slowly", "EOI" };

	isValid = interpreter(shouldFail1);
	if (!isValid) {
		std::cout << "good 1" << std::endl;
	}
	else {
		std::cout << "fail 1" << std::endl;
	}

	isValid = interpreter(shouldFail2);
	if (!isValid) {
		std::cout << "good 2" << std::endl;
	}
	else {
		std::cout << "fail 2" << std::endl;
	}

	isValid = interpreter(shouldFail3);
	if (!isValid) {
		std::cout << "good 3" << std::endl;
	}
	else {
		std::cout << "fail 3" << std::endl;
	}

	isValid = interpreter(shouldFail4);
	if (!isValid) {
		std::cout << "good 4" << std::endl;
	}
	else {
		std::cout << "fail 4" << std::endl;
	}
	isValid = interpreter(shouldFail5);
	if (!isValid) {
		std::cout << "good 1" << std::endl;
	}
	else {
		std::cout << "fail 1" << std::endl;
	}

	isValid = interpreter(shouldFail6);
	if (!isValid) {
		std::cout << "good 2" << std::endl;
	}
	else {
		std::cout << "fail 2" << std::endl;
	}

	isValid = interpreter(shouldFail7);
	if (!isValid) {
		std::cout << "good 3" << std::endl;
	}
	else {
		std::cout << "fail 3" << std::endl;
	}

	isValid = interpreter(shouldFail8);
	if (!isValid) {
		std::cout << "good 4" << std::endl;
	}
	else {
		std::cout << "fail 4" << std::endl;
	}


	return 0;
}